src/Expr.cc

```diff
@@ -3844,49 +3844,36 @@ ArithCoerceExpr::ArithCoerceExpr(ExprPtr arg_op, TypeTag t)
 		ExprError("bad coercion value");
 	}
 
-ValPtr ArithCoerceExpr::FoldSingleVal(Val* v, InternalTypeTag t) const
+ValPtr ArithCoerceExpr::FoldSingleVal(ValPtr v, const TypePtr& t) const
 	{
-	switch ( t ) {
-	case TYPE_INTERNAL_DOUBLE:
-		return make_intrusive<DoubleVal>(v->CoerceToDouble());
-
-	case TYPE_INTERNAL_INT:
-		return val_mgr->Int(v->CoerceToInt());
-
-	case TYPE_INTERNAL_UNSIGNED:
-		return val_mgr->Count(v->CoerceToUnsigned());
-
-	default:
-		RuntimeErrorWithCallStack("bad type in CoerceExpr::Fold");
-		return nullptr;
-	}
+	return check_and_promote(v, t.get(), false, location);
```

Comment:
This isn't really related to this change, but it looks like most calls to `check_and_promote` use `.get()` to pass a raw pointer instead of just passing along the intrusive one. It might be good to fix that at some point.

Reply:
I bundled this in with the updates, too.  I only went so far (the next biggie would be to change `Expr::InitVal`, but that requires touching quite a bit of code).

Reply:
Sounds good. Can you write an issue with what's left so we don't forget about it?

Reply:
Done per https://github.com/zeek/zeek/issues/1624.